File: clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution
{
public:
    Node *cloneGraph(Node *node)
    {
        unordered_map<Node *, Node *> nodeMap;
        if (!node)
            return nullptr;
        return deepCopy(node, nodeMap);
    }
    Node *deepCopy(Node *node, unordered_map<Node *, Node *> &nodeMap)
    {
        if (nodeMap.find(node) != nodeMap.end())
            return nodeMap.at(node);
        Node *newNode = new Node(node->val);
        nodeMap.insert({node, newNode});
        for (Node *&i : node->neighbors)
        {
            newNode->neighbors.push_back(deepCopy(i, nodeMap));
        }
        return newNode;
    }
};

File: clone_graph.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution(object):
    def cloneGraphMem(self, node, visited):
        if node is None:
            return None
        graph = Node(node.val)
        visited[node.val] = graph
        for n in node.neighbors:
            if n.val in visited:
                graph.neighbors.append(visited[n.val])
            else:
                g = self.cloneGraphMem(n, visited)
                graph.neighbors.append(g)
        return graph
        
    def cloneGraph(self, node):
        """
        :type node: Node
        :rtype: Node
        """
        visited = {}
        return self.cloneGraphMem(node, visited)
        


        